src/channel/fs_verifier_channel.rs

```diff
@@ -0,0 +1,106 @@
+use crate::channel::Channel;
+use crate::channel::ChannelStates;
+use crate::hashutil::TempHashContainer;
+use ark_ff::Field;
+use std::marker::PhantomData;
+
+use rand_chacha::rand_core::RngCore;
+use rand_chacha::ChaCha20Rng;
+
+use super::VerifierChannel;
+
+pub struct FSVerifierChannel<F: Field, H: TempHashContainer> {
+    _ph: PhantomData<(F, H)>,
+    prng: ChaCha20Rng,
+    proof: Vec<u8>,
+    proof_read_index: usize,
+    states: ChannelStates,
+}
+
+impl<F: Field, H: TempHashContainer> AsMut<ChannelStates> for FSVerifierChannel<F, H> {
+    fn as_mut(&mut self) -> &mut ChannelStates {
+        &mut self.states
+    }
+}
+
+impl<F: Field, H: TempHashContainer> AsRef<ChannelStates> for FSVerifierChannel<F, H> {
+    fn as_ref(&self) -> &ChannelStates {
+        &self.states
+    }
+}
+
+impl<F: Field, H: TempHashContainer> Channel for FSVerifierChannel<F, H> {
+    type Field = F;
+
+    fn recv_felem(&mut self, felem: Self::Field) -> Result<Self::Field, anyhow::Error> {
+        // TODO
+        Ok(felem)
+    }
+
+    fn recv_bytes(&mut self, n: usize) -> Result<Vec<u8>, anyhow::Error> {
+        assert!(
+            self.proof_read_index + n <= self.proof.len(),
+            "Proof too short."
+        );
+        let raw_bytes = self.proof[self.proof_read_index..self.proof_read_index + n].to_vec();
```

Comment:
You need to add these to the transcript.

Additionally, you can turn the proof into an Iterator<Item = u8> or Reader which avoids you having to manually do the indexing:

https://doc.rust-lang.org/stable/std/io/struct.Cursor.html